Migrate from protoc to rust-protobuf

kckeiks opened the pull request:
# Description

Closes #3024.
Fixes #2645.

## Links to any relevant issues

* https://github.com/stepancheg/rust-protobuf/pull/525

## Open Questions

I updated core and a couple of other packages. This is wip atm but I'll update the remaining crates this week.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

kckeiks committed:
Replace prost with protobuf in misc/prost-codec

kckeiks committed:
Replace prost with protobuf in core

kckeiks committed:
Replace prost with protobuf in transports/plaintext

thomaseizinger commented on core/src/identity.rs:
```diff
@@ -174,32 +176,33 @@ impl Keypair {
         };
 
-        Ok(pk.encode_to_vec())
+        Ok(pk.write_to_bytes().map_err(|e| DecodingError::new("Failed to decode.").source(e))?)
```
When can this fail?

thomaseizinger replied:
Looking at the docs a bit, I think this can only fail on bad usage and since that is static, we should be fine with panicking.

Would you mind investigating this in a bit more detail to confirm? Thank you :)

kckeiks replied:
Yeah I think we are good with panicking https://github.com/stepancheg/rust-protobuf/issues/404

thomaseizinger replied:
That is good to know but also annoying that we need to document it.

It would be nice if they'd implement a trait for all proto3 messages that has an infallible serializer.

thomaseizinger replied:
That is good to know but also annoying that we need to document it.

It would be nice if they'd implement a trait for all proto3 messages that has an infallible serializer.

kckeiks replied:
You beat me to it :blush: https://github.com/stepancheg/rust-protobuf/issues/404#issuecomment-1287460286.

thomaseizinger commented on core/build.rs:
```diff
@@ -19,13 +19,19 @@
-    .unwrap();
+            ]
+        )
+        .customize(cust.gen_mod_rs(true))
```
It says [here](https://docs.rs/protobuf-codegen/latest/protobuf_codegen/struct.Customize.html#method.gen_mod_rs) that this is the default in version 3?

kckeiks replied:
It wasn't working when I was testing it and I don't [see](https://docs.rs/protobuf-codegen/latest/src/protobuf_codegen/customize/mod.rs.html#163-166) how where it defaults to true. I'll test it again.

kckeiks replied:
You were right about this so I'm removing it. I must have overlooked something.

thomaseizinger commented on core/build.rs:
```diff
@@ -19,13 +19,19 @@
 fn main() {
-    prost_build::compile_protos(
-        &[
+    let cust: protobuf_codegen::Customize = Default::default();
```
I think we don't use any of the features described in `lite_runtime`: https://docs.rs/protobuf-codegen/latest/protobuf_codegen/struct.Customize.html#method.lite_runtime

Can you check if we can enable that?

kckeiks replied:
Done.

thomaseizinger commented on core/src/identity.rs:
```diff
@@ -309,37 +316,39 @@ impl TryFrom<keys_proto::PublicKey> for PublicKey {
+        use protobuf::Enum;
+
+        let key_type = keys_proto::KeyType::from_i32(pubkey.Type().value())
+            .ok_or_else(|| DecodingError::new(format!("unknown key type: {}", pubkey.Type().value())))?;
```
Are we roundtripping here? 🤔

kckeiks replied:
Now that I'm looking at it `keys_proto::KeyType::from_i32(pubkey.Type().value()` will not fail. ~I'll add an `expect`.~ I'll do something else :).

thomaseizinger commented on core/src/peer_record.rs:
```diff
@@ -129,7 +126,7 @@ pub enum FromEnvelopeError {
     BadPayload(signed_envelope::ReadPayloadError),
     /// Failed to decode the provided bytes as a [`PeerRecord`].
-    InvalidPeerRecord(prost::DecodeError),
+    InvalidPeerRecord(protobuf::Error),
```
Is this part of the public API? If yes, then this is actually a breaking change which is unfortunate :/

kckeiks replied:
Oh I missed that. Yes, it is part of the public API :disappointed:.

thomaseizinger replied:
Damn okay. Well don't worry about it here for now, I'll send a PR that wraps the error here. That is still a breaking change but we can ship that separately!

thomaseizinger replied:
So my idea here is to make a dedicated error like this:

```rust
#[derive(thiserror::Error, Debug)]
#[error(transparent)]
pub struct DecodeError(protobuf::Error)
```

This preserves the source chain but hides the concrete error type from the public API.

It'd be nice to do this in a separate PR and then put this feature on top of that. Let me know if you'd be happy to do that otherwise I'll try to get to that next week!

kckeiks replied:
> So my idea here is to make a dedicated error like this:
> 
> ```rust
> #[derive(thiserror::Error, Debug)]
> #[error(transparent)]
> pub struct DecodeError(protobuf::Error)
> ```
> 
> This preserves the source chain but hides the concrete error type from the public API.

That sounds good! 

> It'd be nice to do this in a separate PR and then put this feature on top of that. Let me know if you'd be happy to do that otherwise I'll try to get to that next week!

If you could please do that that would be great! I'll rebase once that is merged.

thomaseizinger replied:
I starting working on this here: https://github.com/libp2p/rust-libp2p/pull/3058

thomaseizinger reviewed (commented):
Thank you for tackling this! Very appreciated!

I left a few questions.

kckeiks committed:
Replace prost with protobuf in protocols/autonat

kckeiks committed:
Use lite runtime

kckeiks committed:
Remove roundtripping

kckeiks committed:
Replace prost with protobuf in protocols/dcutr

thomaseizinger commented on core/build.rs:
```diff
@@ -19,13 +19,19 @@
-    .unwrap();
+            ]
+        )
+        .customize(cust.lite_runtime(true))
```
All the builder fns take `self` so I think we can just do this:

```suggestion
        .customize(protobuf_codegen::Customize::default().lite_runtime(true))
```

thomaseizinger commented on core/src/identity.rs:
```diff
@@ -251,23 +254,19 @@ impl PublicKey {
-            .encode(&mut buf)
-            .expect("Vec<u8> provides capacity as needed");
-        buf
+        public_key.write_to_bytes().expect("Encoding failed.")
```
Can we say something like "all fields to be initialized" given that's the invariant we are relying on?

thomaseizinger commented on protocols/autonat/src/protocol.rs:
```diff
@@ -269,28 +266,30 @@ impl DialResponse {
+        let mut msg = structs_proto::Message::new();
+        msg.set_type(structs_proto::message::MessageType::DIAL_RESPONSE as _);
+        msg.dial = MessageField::none();
+        msg.dialResponse = MessageField::some(dial_response);
```
Well this is kind of an annoying API but I guess that is what you get with generated code.

kckeiks replied:
> Well this is kind of an annoying API

Agreed. It would be nice if at least it generated setters for `MessageField`s.

thomaseizinger commented on protocols/dcutr/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+# 0.7.1 [unreleased]
+
+- Remove `prost` and add `protobuf`. See [PR 3050].
```
We should mention in all changelogs that this removes the `protoc` dependency.

thomaseizinger replied:
Can you also remove the `protoc` install step from all workflows please?

thomaseizinger commented on transports/plaintext/src/error.rs:
```diff
@@ -28,7 +28,7 @@ pub enum PlainTextError {
 
     /// Failed to parse the handshake protobuf message.
-    InvalidPayload(Option<prost::DecodeError>),
+    InvalidPayload(Option<protobuf::Error>),
```
This makes it a breaking change too I think.

kckeiks replied:
Yes, it's public.

thomaseizinger reviewed (commented):
Thanks!

kckeiks committed:
Replace prost with protobuf in protocols/floodsub

kckeiks committed:
Add expected invariant

The method write_to_bytes only fails when writting a proto2 message with missing required fields.

kckeiks committed:
Replace prost with protobuf in protocols/gossipsub

thomaseizinger commented on misc/prost-codec/Cargo.toml:
```diff
@@ -13,9 +13,6 @@ categories = ["asynchronous"]
 asynchronous-codec = { version = "0.6" }
 bytes = { version = "1" }
-prost = "0.11"
+protobuf = "3.2"
```
This isn't really a `prost-codec` anymore, now that we are replacing the implementation with `protobuf` :sweat_smile: 

@mxinden Should we publish a new crate?

mxinden replied:
:+1:

kckeiks committed:
Replace prost with protobuf in protocols/identify

kckeiks committed:
Replace prost with protobuf in protocols/kad

kckeiks committed:
Replace prost with protobuf in protocols/relay

kckeiks committed:
Replace prost with protobuf in protocols/rendezvous

thomaseizinger commented on protocols/relay/src/v2/protocol/inbound_hop.rs:
```diff
@@ -136,40 +145,36 @@ pub struct ReservationReq {
-        };
+        let mut msg = HopMessage::new();
+        msg.set_type(hop_message::Type::STATUS);
+        msg.set_status(Status::OK);
```
Can we retain the order here please (`status` being set last)? Makes the diff easier to read :)

thomaseizinger commented on protocols/rendezvous/src/codec.rs:
```diff
@@ -385,118 +371,110 @@ impl TryFrom<wire::Message> for Message {
+        let msg_ty = match message.type_ {
+            Some(ty) => ty.enum_value().or(Err(ConversionError::InconsistentWireMessage))?,
+            None => return Err(ConversionError::InconsistentWireMessage),
+        };
```
This can be solved with the `ok_or` combinator on `Option`.

thomaseizinger commented on protocols/rendezvous/src/codec.rs:
```diff
@@ -385,118 +371,110 @@ impl TryFrom<wire::Message> for Message {
+
+        let message = match msg_ty {
+            MessageType::REGISTER => {
+                match message.register.into_option() {
```
I think this may be cleaner too if we'd use `ok_or(ConversionError::InconsistentWireMessage)?` here and unpack the fields one by one instead of pattern matching on it.

thomaseizinger commented on protocols/rendezvous/src/codec.rs:
```diff
@@ -385,118 +371,110 @@ impl TryFrom<wire::Message> for Message {
+                match message.registerResponse.into_option() {
+                    Some(
+                        RegisterResponse { status: Some(status), ttl, .. }
+                    ) if status.value() == 0 => Message::RegisterResponse(Ok(ttl.ok_or(ConversionError::MissingTtl)?)),
```
This `status` is enum as well right? I think we should probably use that instead of the bare number. With `prost`, we couldn't match on that but now we can :)

kckeiks replied:
It's an enum but it's wrapped in a couple of abstractions. Will take a look again.

thomaseizinger reviewed (commented):
Only a few minor comments!

Looking forward to the full solution :)

kckeiks committed:
Replace prost with protobuf in protocols/noise

kckeiks committed:
Fix keys_proto::KeyType variant spelling

kckeiks committed:
Fix tests in libp2p-core

kckeiks committed:
Fix tests in libp2p-gossipsub

kckeiks committed:
Fix clippy errors

kckeiks committed:
Run cargo fmt

kckeiks committed:
Remove reference to prost in Codec docs

thomaseizinger commented on core/src/identity.rs:
```diff
@@ -174,32 +176,33 @@ impl Keypair {
         };
 
-        Ok(pk.encode_to_vec())
+        Ok(pk.write_to_bytes().map_err(|e| DecodingError::new("Failed to decode.").source(e))?)
```
We discovered that this can never fail right? Let's panic here for consistency.

thomaseizinger commented on core/src/identity.rs:
```diff
@@ -174,32 +176,35 @@ impl Keypair {
         };
 
-        Ok(pk.encode_to_vec())
+        pk.write_to_bytes()
+            .map_err(|e| DecodingError::new("Failed to decode.").source(e))
```
Please panic here for consistency.

thomaseizinger commented on core/src/identity.rs:
```diff
@@ -174,32 +176,35 @@ impl Keypair {
-        let key_type = keys_proto::KeyType::from_i32(private_key.r#type).ok_or_else(|| {
-            DecodingError::new(format!("unknown key type: {}", private_key.r#type))
-        })?;
+        let key_type =
+            keys_proto::KeyType::from_i32(private_key.Type().value()).ok_or_else(|| {
+                DecodingError::new(format!("unknown key type: {:?}", private_key.Type()))
+            })?;
```
This looks like another round-trip to me? Wouldn't `Type()` already give you the enum value?

thomaseizinger commented on core/src/peer_record.rs:
```diff
@@ -30,11 +30,12 @@ impl PeerRecord {
             envelope.payload_and_signing_key(String::from(DOMAIN_SEP), PAYLOAD_TYPE.as_bytes())?;
-        let record = peer_record_proto::PeerRecord::decode(payload)?;
+        let record = peer_record_proto::PeerRecord::parse_from_bytes(payload)
+            .map_err(FromEnvelopeError::from)?;
```
Why do we need the `map_err` now?

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -729,13 +733,16 @@ where
-                m.msg_sent(&topic_hash, msg_bytes);
+                m.msg_sent(
+                    &topic_hash,
+                    usize::try_from(msg_bytes).expect("Size of sent messages fit in usize."),
```
If we panic here, we might as well also panic on line 621.

thomaseizinger replied:
I think as regular cast could do as well.

thomaseizinger reviewed (changes requested):
Thanks!

A few more comments, plus the CI workflow needs adjusting to remove the installation of `protoc`!
